classes/local/store/s3/client.php

```diff
@@ -430,10 +453,103 @@ public function generate_presigned_url($contenthash, $headers) {
         if ($expirationtime !== '' and strtotime($expirationtime) != 0) {
             $request = $this->client->createPresignedRequest($command, $expirationtime);
         } else {
-            $request = $this->client->createPresignedRequest($command, '+'.$this->expirationtime.' seconds');
+            $request = $this->client->createPresignedRequest($command, '+' . $this->expirationtime . ' seconds');
         }
 
         $signedurl = (string)$request->getUri();
+
         return $signedurl;
     }
+
+    /**
+     * Generates pre-signed URL to CDN from its hash.
+     *
+     * @param string $contenthash file content hash.
+     * @param array $headers request headers.
+     * @param bool $nicefilename deliver original filename rather than origin hashed filename.
+     *
+     * @return string.
+     */
+    private function generate_presigned_cloudfront_url($contenthash = '', $headers = array(), $nicefilename = true) {
+
+        if ($this->signingmethod != 'CF') {
+            throw new \Exception('Cloudfront signing method is not enabled.');
+        }
+
+        $cdnconfig = get_objectfs_config();
+        $key = $this->get_filepath_from_hash($contenthash);
+        $cloudfrontclient = new CloudFrontClient(
+            array(
+                'profile' => 'default',
+                'version' => 'latest', /* '2014-11-06' */
+                'region' => $cdnconfig->s3_region,  /* The region is the source bucket region ? - 'ap-southeast-2' */
+            )
+        );
+
+        $resourcedomain = $cdnconfig->cloudfront_resource_domain;
+
+        $resourcekey = $resourcedomain . '/' . $key;
+
+        if ($nicefilename) {
+            // We are trying to deliver original filename rather than hash filename to client.
+
+            $contentdisposition = '';
+            $originalfilename = '';
+            $originalcontenttype = '';
+
+            $contentdisposition = $this->get_header($headers, 'Content-Disposition');
```

Comment:
I think these whole 4-5 lines can be much simpler as just:

$contentdisposition = trim($this->get_header($headers, 'Content-Disposition'));

I think the if is redundant. Same might apply to contenttype